lightning/src/ln/onion_utils.rs

```diff
@@ -381,224 +396,242 @@ pub(super) fn build_first_hop_failure_packet(shared_secret: &[u8], failure_type:
-							network_update = Some(NetworkUpdate::NodeFailure {
-								node_id: route_hop.pubkey,
-								is_permanent: true,
-							});
-							short_channel_id = Some(route_hop.short_channel_id);
-						}
-
-						res = Some((network_update, short_channel_id, !(error_code & PERM == PERM && is_from_final_node)));
-
-						let (description, title) = errors::get_onion_error_description(error_code);
-						if debug_field_size > 0 && err_packet.failuremsg.len() >= 4 + debug_field_size {
-							log_info!(logger, "Onion Error[from {}: {}({:#x}) {}({})] {}", route_hop.pubkey, title, error_code, debug_field, log_bytes!(&err_packet.failuremsg[4..4+debug_field_size]), description);
-						}
-						else {
-							log_info!(logger, "Onion Error[from {}: {}({:#x})] {}", route_hop.pubkey, title, error_code, description);
-						}
-					} else {
-						// Useless packet that we can't use but it passed HMAC, so it
-						// definitely came from the peer in question
-						let network_update = Some(NetworkUpdate::NodeFailure {
-							node_id: route_hop.pubkey,
-							is_permanent: true,
-						});
-						let short_channel_id = Some(route_hop.short_channel_id);
-						res = Some((network_update, short_channel_id, !is_from_final_node));
-					}
+		(path, session_priv, first_hop_htlc_msat)
+	} else { unreachable!() };
+	let mut res = None;
+	let mut htlc_msat = *first_hop_htlc_msat;
+	let mut error_code_ret = None;
+	let mut error_packet_ret = None;
+	let mut is_from_final_node = false;
+
+	// Handle packed channel/node updates for passing back for the route handler
+	construct_onion_keys_callback(secp_ctx, &path, session_priv,
+		|shared_secret, _, pubkey, route_hop_opt, route_hop_idx|
+	{
+		if res.is_some() { return; }
+
+		let ammag = gen_ammag_from_shared_secret(shared_secret.as_ref());
+
+		let mut decryption_tmp = Vec::with_capacity(packet_decrypted.len());
+		decryption_tmp.resize(packet_decrypted.len(), 0);
+		let mut chacha = ChaCha20::new(&ammag, &[0u8; 8]);
+		chacha.process(&packet_decrypted, &mut decryption_tmp[..]);
+		packet_decrypted = decryption_tmp;
+
+		let err_packet = match msgs::DecodedOnionErrorPacket::read(&mut Cursor::new(&packet_decrypted)) {
+			Ok(p) => p,
+			Err(_) => return
+		};
+		let um = gen_um_from_shared_secret(shared_secret.as_ref());
+		let mut hmac = HmacEngine::<Sha256>::new(&um);
+		hmac.input(&err_packet.encode()[32..]);
+
+		if !fixed_time_eq(&Hmac::from_engine(hmac).into_inner(), &err_packet.hmac) { return }
+		let error_code_slice = match err_packet.failuremsg.get(0..2) {
+			Some(s) => s,
+			None if route_hop_opt.is_some() => {
+				// Useless packet that we can't use but it passed HMAC, so it definitely came from the peer
+				// in question
+				let network_update = Some(NetworkUpdate::NodeFailure {
+					node_id: pubkey,
+					is_permanent: true,
+				});
+				let short_channel_id = Some(route_hop_opt.as_ref().unwrap().short_channel_id);
+				res = Some((network_update, short_channel_id, !is_from_final_node));
```

Comment:
Where we set `is_from_final_node` got moved from before to after this section which seems it may end up returning the wrong value here in some cases?

Reply:
Good point, we had preexisting missing test coverage here. Added now!